src/core/bitset.c

```diff
@@ -0,0 +1,648 @@
+ * bitset. Let us see a small example to clarify things: if you
+ * reserve space for 100 elements and the size of your
+ * bitset was (and still is) 60, then you can surely add additional 40
+ * elements to your bitset before it will be copied.
+ * \param bitset The bitset object.
+ * \param capacity The new \em allocated size of the bitset.
+ * \return Error code:
+ *         \c IGRAPH_ENOMEM if there is not enough memory.
+ *
+ * Time complexity: operating system dependent, should be around
+ * O(n/w),
+ * n is the new allocated size of the bitset,
+ * w is the word size of the machine (32 or 64).
+ */
+
+igraph_error_t igraph_bitset_reserve(igraph_bitset_t *bitset, igraph_integer_t capacity) {
+    igraph_integer_t current_capacity;
+    igraph_integer_t *tmp;
+
+    IGRAPH_ASSERT(bitset != NULL);
+    IGRAPH_ASSERT(bitset->stor_begin != NULL);
+    IGRAPH_ASSERT(capacity >= 0);
+
+    current_capacity = igraph_bitset_capacity(bitset);
+
+    if (IGRAPH_BIT_NSLOTS(capacity) <= IGRAPH_BIT_NSLOTS(current_capacity)) {
+        return IGRAPH_SUCCESS;
+    }
+
+    tmp = IGRAPH_REALLOC(bitset->stor_begin, IGRAPH_BIT_NSLOTS(capacity), igraph_integer_t);
+    IGRAPH_CHECK_OOM(tmp, "Cannot reserve space for bitset.");
+
+    bitset->stor_begin = tmp;
+    bitset->stor_end = bitset->stor_begin + IGRAPH_BIT_NSLOTS(capacity);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_resize
+ * \brief Resizes the bitset.
+ *
+ * \experimental
+ *
+ * </para><para>
+ * Note that this function does not free any memory, just sets the
+ * size of the bitset to the given one. It may, on the other hand,
+ * allocate more memory if the new size is larger than the previous
+ * one. In this case the newly appeared elements in the bitset are
+ * \em not set to zero, they are uninitialized.
+ *
+ * \param bitset The bitset object
+ * \param new_size The new size of the bitset.
+ * \return Error code,
+ *         \c IGRAPH_ENOMEM if there is not enough
+ *         memory. Note that this function \em never returns an error
+ *         if the bitset is made smaller.
+ * \sa \ref igraph_bitset_reserve() for allocating memory for future
+ * extensions of a bitset.
+ *
+ * Time complexity: O(1) if the new
+ * size is smaller, operating system dependent if it is larger. In the
+ * latter case it is usually around
+ * O(n/w),
+ * n is the new size of the bitset,
+ * w is the word size of the machine (32 or 64).
+ */
+
+igraph_error_t igraph_bitset_resize(igraph_bitset_t *bitset, igraph_integer_t new_size) {
+    IGRAPH_ASSERT(bitset != NULL);
+    IGRAPH_ASSERT(bitset->stor_begin != NULL);
+    IGRAPH_CHECK(igraph_bitset_reserve(bitset, new_size));
+
+    for (igraph_integer_t i = bitset->size; i % IGRAPH_INTEGER_SIZE != 0; ++i) {
+        IGRAPH_BIT_CLEAR(*bitset, i);
+    }
+    for (igraph_integer_t i = IGRAPH_BIT_NSLOTS(bitset->size); i < IGRAPH_BIT_NSLOTS(new_size); ++i) {
+        VECTOR(*bitset)[i] = 0;
+    }
```

Comment:
This could be done more efficiently with a `memset()`

Reply:
Updated